### ISSUE TITLE
security group not idempotent

### ISSUE THREAD

**Issue by developerinlondon** [hashicorp/terraform#7680]
hi there,

i am finding that terraform is not idempotent when creating security group with an ipv4 4 protocol.

the below can be tried to reproduce the issue:
```

resource "aws_security_group" "vpcipv4" {
  name   = "ipv4vpc"

  tags {
    Name = "ipv4vpc"
  }

  ingress {
    from_port = 0
    to_port = 65535
    protocol = "4"
    cidr_blocks = ["0.0.0.0/0"]
  }
  egress {
    from_port = 0
    to_port = 65535
    protocol = "4"
    cidr_blocks = ["0.0.0.0/0"]
  }

}
```

now you can run:
```
terraform plan
terraform apply
terraform plan
```
and it would show it needs to be updated again.

let me know what i can do to fix this issue.

thanks.

**Comment by ketzacoatl** [comment]
Can you paste the contents of the output you seen when you apply the plan and then review it again?

**Comment by stack72** [comment]
I was able to recreate this:

```
~ aws_security_group.vpcipv4
    egress.2122438593.cidr_blocks.#:      "0" => "1"
    egress.2122438593.cidr_blocks.0:      "" => "0.0.0.0/0"
    egress.2122438593.from_port:          "" => "0"
    egress.2122438593.prefix_list_ids.#:  "0" => "0"
    egress.2122438593.protocol:           "" => "4"
    egress.2122438593.security_groups.#:  "0" => "0"
    egress.2122438593.self:               "" => "false"
    egress.2122438593.to_port:            "" => "65535"
    egress.3485474627.cidr_blocks.#:      "1" => "0"
    egress.3485474627.cidr_blocks.0:      "0.0.0.0/0" => ""
    egress.3485474627.from_port:          "0" => "0"
    egress.3485474627.prefix_list_ids.#:  "0" => "0"
    egress.3485474627.protocol:           "4" => ""
    egress.3485474627.security_groups.#:  "0" => "0"
    egress.3485474627.self:               "false" => "false"
    egress.3485474627.to_port:            "0" => "0"
    ingress.2122438593.cidr_blocks.#:     "0" => "1"
    ingress.2122438593.cidr_blocks.0:     "" => "0.0.0.0/0"
    ingress.2122438593.from_port:         "" => "0"
    ingress.2122438593.protocol:          "" => "4"
    ingress.2122438593.security_groups.#: "0" => "0"
    ingress.2122438593.self:              "" => "false"
    ingress.2122438593.to_port:           "" => "65535"
    ingress.3485474627.cidr_blocks.#:     "1" => "0"
    ingress.3485474627.cidr_blocks.0:     "0.0.0.0/0" => ""
    ingress.3485474627.from_port:         "0" => "0"
    ingress.3485474627.protocol:          "4" => ""
    ingress.3485474627.security_groups.#: "0" => "0"
    ingress.3485474627.self:              "false" => "false"
    ingress.3485474627.to_port:           "0" => "0"
```

That's the continual plan

Paul

**Comment by apparentlymart** [comment]
Looking closely at this diff, it seems that the only difference here is that `to_port` is changing from zero to 65535 for both `egress` and `ingress`.

This looks like another example of the classic problem of AWS doing undocumented normalization of the data when it's stored and then retrieved from the API. In that case, the workaround would be to change the configuration so that `to_port` is explicitly 65535, and then AWS should return it back the same way as it's written in the config, preventing this perpetual diff.

**Comment by ketzacoatl** [comment]
has this been resolved in 0.6.x?
